internal/proxy/cache.go

```diff
@@ -0,0 +1,205 @@
+package proxy
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/vault/sdk/helper/cryptoutil"
+	"github.com/hashicorp/vault/sdk/helper/strutil"
+	gocache "github.com/patrickmn/go-cache"
+)
+
+const (
+	vaultCacheTTL = "VAULT_DEFAULT_CACHE_TTL"
+
+	// Return a cached response if it exists, otherwise fall back to Vault and
+	// cache the response
+	headerOptionCacheable = "cache"
+
+	// Send the request to Vault and cache the response
+	headerOptionRecache = "recache"
+)
+
+type Cache struct {
+	data    *gocache.Cache
+	timeout time.Duration
+}
+
+type CacheKey struct {
+	Token       string
+	Request     *http.Request
+	RequestBody []byte
+}
+
+type CacheData struct {
+	StatusCode int
+	Header     http.Header
+	Body       []byte
+}
+
+type CacheOptions struct {
+	cacheable bool
+	recache   bool
+}
+
+func NewCache(timeout time.Duration) *Cache {
+	return &Cache{
+		data:    gocache.New(timeout, timeout),
+		timeout: timeout,
+	}
+}
+
+// constructs the CacheKey for this request and token and returns the SHA256
+// hash
+func makeRequestHash(logger *log.Logger, r *http.Request, token string) (string, error) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read request body: %w", err)
+	}
+	if r.Body != nil {
+		r.Body.Close()
```

Comment:
Is it possible that the body also needs closing when there is an error above? And should we also log the returned error from `Close` if it's non-nil?

Reply:
I'm not really sure the body needs closing when there is an error above, since an error here will stop processing the request in the calling handler. That being said I added more logging and another call to r.Body.Close() out of an abundance of caution in 73bfb8c815700f7a4bbe6951eae49d7c95579244.

Reply:
Yeah sorry, to be clear, I actually wasn't sure of the answers to my questions. As you say, handling those cases is showing an abundance of caution, but LGTM 👍